File: csharp/Platform.Protocols.Lino.Tests/MixedIndentationModesTests.cs
using System;
using Xunit;

namespace Platform.Protocols.Lino.Tests
{
    public static class MixedIndentationModesTests
    {
        [Fact]
        public static void HeroExampleMixedModesTest()
        {
            var input = @"empInfo
  employees:
    (
      name (James Kirk)
      age 40
    )
    (
      name (Jean-Luc Picard)
      age 45
    )
    (
      name (Wesley Crusher)
      age 27
    )";

            var parser = new Parser();
            var result = parser.Parse(input);

            Assert.NotEmpty(result);
            var formatted = result.Format();
            Assert.Contains("empInfo", formatted);
            Assert.Contains("employees", formatted);
            Assert.Contains("James Kirk", formatted);
            Assert.Contains("Jean-Luc Picard", formatted);
            Assert.Contains("Wesley Crusher", formatted);
        }

        [Fact]
        public static void HeroExampleAlternativeFormatTest()
        {
            var input = @"empInfo
  (
    employees:
      (
        name (James Kirk)
        age 40
      )
      (
        name (Jean-Luc Picard)
        age 45
      )
      (
        name (Wesley Crusher)
        age 27
      )
  )";

            var parser = new Parser();
            var result = parser.Parse(input);

            Assert.NotEmpty(result);
            var formatted = result.Format();
            Assert.Contains("empInfo", formatted);
            Assert.Contains("employees:", formatted);
            Assert.Contains("James Kirk", formatted);
            Assert.Contains("Jean-Luc Picard", formatted);
            Assert.Contains("Wesley Crusher", formatted);
        }

        [Fact]
        public static void HeroExampleEquivalenceTest()
        {
            var version1 = @"empInfo
  employees:
    (
      name (James Kirk)
      age 40
    )
    (
      name (Jean-Luc Picard)
      age 45
    )
    (
      name (Wesley Crusher)
      age 27
    )";

            var version2 = @"empInfo
  (
    employees:
      (
        name (James Kirk)
        age 40
      )
      (
        name (Jean-Luc Picard)
        age 45
      )
      (
        name (Wesley Crusher)
        age 27
      )
  )";

            var parser = new Parser();
            var result1 = parser.Parse(version1);
            var result2 = parser.Parse(version2);

            Assert.NotEmpty(result1);
            Assert.NotEmpty(result2);

            var formatted1 = result1.Format();
            var formatted2 = result2.Format();

            Assert.Equal(formatted1, formatted2);
        }

        [Fact]
        public static void SetContextWithoutColonTest()
        {
            var input = @"empInfo
  employees";

            var parser = new Parser();
            var result = parser.Parse(input);

            Assert.NotEmpty(result);
            var formatted = result.Format();
            Assert.Contains("empInfo", formatted);
            Assert.Contains("employees", formatted);
        }

        [Fact]
        public static void SequenceContextWithColonTest()
        {
            var input = @"employees:
  James Kirk
  Jean-Luc Picard
  Wesley Crusher";

            var parser = new Parser();
            var result = parser.Parse(input);

            Assert.NotEmpty(result);
            Assert.Single(result);
            var formatted = result.Format();
            Assert.Contains("employees:", formatted);
            Assert.Contains("James Kirk", formatted);
            Assert.Contains("Jean-Luc Picard", formatted);
            Assert.Contains("Wesley Crusher", formatted);
        }

        [Fact]
        public static void SequenceContextWithComplexValuesTest()
        {
            var input = @"employees:
  (
    name (James Kirk)
    age 40
  )
  (
    name (Jean-Luc Picard)
    age 45
  )";

            var parser = new Parser();
            var result = parser.Parse(input);

            Assert.NotEmpty(result);
            Assert.Single(result);
            var formatted = result.Format();
            Assert.Contains("employees:", formatted);
            Assert.Contains("James Kirk", formatted);
            Assert.Contains("Jean-Luc Picard", formatted);
        }

        [Fact]
        public static void NestedSetAndSequenceContextsTest()
        {
            var input = @"company
  departments:
    engineering
    sales
  employees:
    (name John)
    (name Jane)";

            var parser = new Parser();
            var result = parser.Parse(input);

            Assert.NotEmpty(result);
            var formatted = result.Format();
            Assert.Contains("company", formatted);
            Assert.Contains("departments:", formatted);
            Assert.Contains("employees:", formatted);
        }

        [Fact]
        public static void DeeplyNestedMixedModesTest()
        {
            var input = @"root
  level1
    level2:
      value1
      value2
    level2b
      level3";

            var parser = new Parser();
            var result = parser.Parse(input);

            Assert.NotEmpty(result);
            var formatted = result.Format();
            Assert.Contains("root", formatted);
            Assert.Contains("level2:", formatted);
        }
    }
}

File: rust/tests/mixed_indentation_modes_tests.rs
use links_notation::parse_lino_to_links;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hero_example_mixed_modes_test() {
        let input = "empInfo\n  employees:\n    (\n      name (James Kirk)\n      age 40\n    )\n    (\n      name (Jean-Luc Picard)\n      age 45\n    )\n    (\n      name (Wesley Crusher)\n      age 27\n    )";

        let result = parse_lino_to_links(input).unwrap();

        assert!(!result.is_empty());
        let formatted = result.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        assert!(formatted.contains("empInfo"));
        assert!(formatted.contains("employees:"));
        assert!(formatted.contains("James Kirk") || formatted.contains("James") && formatted.contains("Kirk"));
        assert!(formatted.contains("Jean-Luc Picard") || formatted.contains("Jean-Luc") && formatted.contains("Picard"));
        assert!(formatted.contains("Wesley Crusher") || formatted.contains("Wesley") && formatted.contains("Crusher"));
    }

    #[test]
    fn hero_example_alternative_format_test() {
        let input = "empInfo\n  (\n    employees:\n      (\n        name (James Kirk)\n        age 40\n      )\n      (\n        name (Jean-Luc Picard)\n        age 45\n      )\n      (\n        name (Wesley Crusher)\n        age 27\n      )\n  )";

        let result = parse_lino_to_links(input).unwrap();

        assert!(!result.is_empty());
        let formatted = result.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        assert!(formatted.contains("empInfo"));
        assert!(formatted.contains("employees:"));
        assert!(formatted.contains("James Kirk") || formatted.contains("James") && formatted.contains("Kirk"));
        assert!(formatted.contains("Jean-Luc Picard") || formatted.contains("Jean-Luc") && formatted.contains("Picard"));
        assert!(formatted.contains("Wesley Crusher") || formatted.contains("Wesley") && formatted.contains("Crusher"));
    }

    #[test]
    fn hero_example_equivalence_test() {
        let version1 = "empInfo\n  employees:\n    (\n      name (James Kirk)\n      age 40\n    )\n    (\n      name (Jean-Luc Picard)\n      age 45\n    )\n    (\n      name (Wesley Crusher)\n      age 27\n    )";

        let version2 = "empInfo\n  (\n    employees:\n      (\n        name (James Kirk)\n        age 40\n      )\n      (\n        name (Jean-Luc Picard)\n        age 45\n      )\n      (\n        name (Wesley Crusher)\n        age 27\n      )\n  )";

        let result1 = parse_lino_to_links(version1).unwrap();
        let result2 = parse_lino_to_links(version2).unwrap();

        let formatted1 = result1.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        let formatted2 = result2.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");

        assert_eq!(formatted1, formatted2);
    }

    #[test]
    fn set_context_without_colon_test() {
        let input = "empInfo\n  employees";

        let result = parse_lino_to_links(input).unwrap();

        assert!(!result.is_empty());
        let formatted = result.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        assert!(formatted.contains("empInfo"));
        assert!(formatted.contains("employees"));
    }

    #[test]
    fn sequence_context_with_colon_test() {
        let input = "employees:\n  James Kirk\n  Jean-Luc Picard\n  Wesley Crusher";

        let result = parse_lino_to_links(input).unwrap();

        assert!(!result.is_empty());
        assert_eq!(result.len(), 1);
        let formatted = result.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        assert!(formatted.contains("employees:"));
        assert!(formatted.contains("James Kirk") || formatted.contains("James") && formatted.contains("Kirk"));
        assert!(formatted.contains("Jean-Luc Picard") || formatted.contains("Jean-Luc") && formatted.contains("Picard"));
        assert!(formatted.contains("Wesley Crusher") || formatted.contains("Wesley") && formatted.contains("Crusher"));
    }

    #[test]
    fn sequence_context_with_complex_values_test() {
        let input = "employees:\n  (\n    name (James Kirk)\n    age 40\n  )\n  (\n    name (Jean-Luc Picard)\n    age 45\n  )";

        let result = parse_lino_to_links(input).unwrap();

        assert!(!result.is_empty());
        assert_eq!(result.len(), 1);
        let formatted = result.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        assert!(formatted.contains("employees:"));
        assert!(formatted.contains("James Kirk") || formatted.contains("James") && formatted.contains("Kirk"));
        assert!(formatted.contains("Jean-Luc Picard") || formatted.contains("Jean-Luc") && formatted.contains("Picard"));
    }

    #[test]
    fn nested_set_and_sequence_contexts_test() {
        let input = "company\n  departments:\n    engineering\n    sales\n  employees:\n    (name John)\n    (name Jane)";

        let result = parse_lino_to_links(input).unwrap();

        assert!(!result.is_empty());
        let formatted = result.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        assert!(formatted.contains("company"));
        assert!(formatted.contains("departments:"));
        assert!(formatted.contains("employees:"));
    }

    #[test]
    fn deeply_nested_mixed_modes_test() {
        let input = "root\n  level1\n    level2:\n      value1\n      value2\n    level2b\n      level3";

        let result = parse_lino_to_links(input).unwrap();

        assert!(!result.is_empty());
        let formatted = result.iter().map(|l| format!("{}", l)).collect::<Vec<_>>().join("\n");
        assert!(formatted.contains("root"));
        assert!(formatted.contains("level2:"));
    }
}
